prototype for SE-0174

Dante-Broggi opened the pull request:
<!-- What's in this pull request? -->
I noticed that despite being accepted since May 6, 2017, over a year ago, there is not even a PR to implement SE-0174, so it being a completely standard library change, I made a prototype.

Note this does not include:
* tests
* source compatibility aids
And also includes:
* custom `Filtered`s for 'obvious' types.
* `where` clauses not specified by the proposal (because it was accepted before `where` clauses)

Due to the age of SE-0174, I believe this certainly needs core team approval, but I do not believe it needs a re-review.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -→
Implements half of [SR-3444](https://bugs.swift.org/browse/SR-3444).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Dante-Broggi committed:
Prototype SE-0174

Dante-Broggi committed:
Remaining Protocols

Dante-Broggi committed:
Concrete types

slavapestov commented on stdlib/public/core/CollectionOfOne.swift:
```diff
@@ -75,6 +75,9 @@ extension CollectionOfOne: RandomAccessCollection, MutableCollection {
   public typealias Indices = Range<Int>
   public typealias SubSequence = Slice<CollectionOfOne<Element>>
+  // FIXME: do we want a special type for this?
+  // a `CollectionOfZeroOrOne`?
```
It's called `Optional`, but unfortunately it does not conform to Collection ;-)

Azoy commented on stdlib/public/core/RandomAccessCollection.swift:
```diff
@@ -152,7 +155,9 @@ where SubSequence: RandomAccessCollection, Indices: RandomAccessCollection
-  /// - Complexity: O(1)
+  /// - Complexity: O(1) if the collection conforms to
+  ///   `RandomAccessCollection`; otherwise, O(*k*), where *k* is the absolute
+  ///   value of `distance`.
```
Didn't you pr the removal of `if the collections conforms to...`?

Dante-Broggi replied:
I did this edit before that (timeline doesn't say so, because I had trouble with Github Desktop not wanting to create a branch on my fork.
Will re-fix.

Dante-Broggi committed:
Re do O(1) change.

I had problems with Github Desktop, and forgot to check after re-doing the changes.

airspeedswift commented:
Hi @Dante-Broggi. Thanks for picking up on this.

So, the reason it wasn't done originally (and this was back in the day before we required implementations before proposals :) is because it turned out we needed recursive protocol constraints i.e. `Sequence.Filtered: Sequence`.

Now having had a year to reflect on things, I kind of feel like I wouldn't propose the associated type if I were writing the proposal today. `RanceReplaceableCollection` being able to return `Self` covers most of the important cases. That leaves `Dictionary` and `Set`, and I'm not sure it's necessary to be able to rely on getting back one of these _in a generic context_. They have their own concrete implementations and maybe that's enough. I'm not sure. Probably needs some discussion. 

In the mean-time I'll kick off a compiler performance run as that's my other worry.

airspeedswift commented:
@swift-ci please test source compatibility

airspeedswift commented:
@swift-ci please smoke benchmark

airspeedswift commented:
@swift-ci please smoke test compiler performance

airspeedswift commented:
@swift-ci please test

Dante-Broggi committed:
Fix missing `[Element]` => `Filtered`

Also some comments, a missing "public", and try something for `_SequenceWrapper`

xwu commented on stdlib/public/core/EmptyCollection.swift:
```diff
@@ -40,11 +40,29 @@ extension EmptyCollection {
+  @inlinable
+  public func filter(
+    _ isIncluded: (Element) throws -> Bool
+    ) rethrows -> Filtered {
```
Nit: standard library style is to unindent this parenthesis.

xwu commented on stdlib/public/core/EmptyCollection.swift:
```diff
@@ -141,6 +160,22 @@ extension EmptyCollection: RandomAccessCollection, MutableCollection {
+  @inlinable
+  public func filter(
+    _ isIncluded: (Element) throws -> Bool
+    ) rethrows -> Filtered {
```
....also here.

xwu commented on stdlib/public/core/Collection.swift:
```diff
@@ -392,6 +392,14 @@ public protocol Collection: Sequence where SubSequence: Collection {
   /// collection, the subsequence should also conform to `Collection`.
   associatedtype SubSequence = Slice<Self> where SubSequence.Index == Index
 
+  /// A type that represents a collection composed of an arbatrary
```
Spelling: arbitrary.

xwu commented on stdlib/public/core/RangeReplaceableCollection.swift:
```diff
@@ -65,6 +65,15 @@ public protocol RangeReplaceableCollection : Collection
   // FIXME(ABI): Associated type inference requires this.
   associatedtype SubSequence
+  
+  /// A type that represents a collection composed of an arbatrary
```
Spelling: arbitrary.

xwu commented on stdlib/public/core/Sequence.swift:
```diff
@@ -330,10 +330,16 @@ public protocol Sequence {
     where Element == SubSequence.Element,
           SubSequence.SubSequence == SubSequence
+  
+  /// A type that represents a sequence composed of an arbatrary
```
Spelling: arbitrary.

xwu commented on stdlib/public/core/Sequence.swift:
```diff
@@ -385,12 +391,12 @@ public protocol Sequence {
   ///   sequence as its argument and returns a Boolean value indicating
   ///   whether the element should be included in the returned array.
-  /// - Returns: An array of the elements that `isIncluded` allowed.
+  /// - Returns: An sequence of the elements that `isIncluded` allowed.
```
Spelling: A sequence.

xwu commented on stdlib/public/core/Sequence.swift:
```diff
@@ -1051,6 +1015,99 @@ extension Sequence where Element : Equatable {
+  @inlinable
+  public func filter(
+    _ isIncluded: (Element) throws -> Bool
+    ) rethrows -> Filtered {
```
...here also.

xwu commented on stdlib/public/core/Sequence.swift:
```diff
@@ -1051,6 +1015,99 @@ extension Sequence where Element : Equatable {
+  @_transparent
+  public func _filter(
+    _ isIncluded: (Element) throws -> Bool
+    ) rethrows -> Filtered {
```
...and here.

xwu commented on stdlib/public/core/Sequence.swift:
```diff
@@ -1051,6 +1015,99 @@ extension Sequence where Element : Equatable {
+  @inlinable
+  public func filter(
+    _ isIncluded: (Element) throws -> Bool
+    ) rethrows -> [Element] {
```
...and here.

xwu commented on stdlib/public/core/Sequence.swift:
```diff
@@ -1051,6 +1015,99 @@ extension Sequence where Element : Equatable {
+  @_transparent
+  public func _filter(
+    _ isIncluded: (Element) throws -> Bool
+    ) rethrows -> [Element] {
```
...and here.

xwu reviewed (commented):
Some nitpicking, in case it's decided to proceed :)

Dante-Broggi committed:
Fix typos / style errors found by xwu

natecook1000 commented:
Closing this PR as it's out of date and a bit dubious that we will merge. @Dante-Broggi, if you'd like to continue discussion on this topic, please open a thread on the forums. Thanks!
